mongoimport/common.go

```diff
@@ -399,28 +570,197 @@ func validateFields(fields []string) error {
+	if len(fieldParts[partIndex:]) == 1 {
+		tree[fieldParts[partIndex]] = true
+		return tree, nil
+	}
+
+	// Now determine the type of the next part of the field
+	partIndex++
+	if _, ok := isNatNum(fieldParts[partIndex]); useArrayIndexFields && ok {
+		var subArray []interface{}
+		if exists {
+			subArray, ok = value.([]interface{})
+			if !ok {
+				// case (5) or (4)
+				return nil, fmt.Errorf("fields '%v' and '%v' are incompatible",
+					strings.Join(fieldParts[:partIndex], ".")+findFirstField(value),
+					strings.Join(fieldParts, "."))
 			}
-			// NOTE: this means we will not support imports that have fields like
-			// a, a - since this is invalid in MongoDB
-			if field == latterField {
-				return fmt.Errorf("fields cannot be identical: '%v' and '%v'", field, latterField)
+		} else {
+			subArray = make([]interface{}, 0)
+		}
+		subArray, err := addFieldToArray(fieldParts, partIndex, subArray)
+		if err != nil {
+			return nil, err
+		}
+		tree[fieldParts[0]] = subArray
+	} else {
+		var subTree map[string]interface{}
+		if exists {
+			subTree, ok = value.(map[string]interface{})
+			if !ok {
+				// case (5) or (4)
+				return nil, fmt.Errorf("fields '%v' and '%v' are incompatible",
+					strings.Join(fieldParts[:partIndex], ".")+findFirstField(value),
+					strings.Join(fieldParts, "."))
 			}
+		} else {
+			subTree = make(map[string]interface{})
+		}
+		subTree, err := addFieldToTree(fieldParts, partIndex, subTree, useArrayIndexFields)
+		if err != nil {
+			return nil, err
 		}
+		tree[fieldParts[0]] = subTree
 	}
-	return nil
+	return tree, nil
+
+}
+
+// addFieldToArray is used with addFieldToTree() to build a valid tree of fields.
+func addFieldToArray(fieldParts []string, partIndex int, array []interface{}) ([]interface{}, error) {
+	// The first part of the field should be an index of an array
+	idx, ok := isNatNum(fieldParts[partIndex])
+	if !ok {
+		// We shouldn't ever get here
+		return nil, fmt.Errorf("addFieldToArray expected an integer field, but instead received %s", fieldParts[0])
+	}
+
+	if len(fieldParts[partIndex:]) == 1 {
+		// We're at the terminus of a field so we have to check if we can append to the array
+		if idx < len(array) {
+			// case (2)
+			return nil, fmt.Errorf("fields cannot be identical: '%v' and '%v'",
+				strings.Join(fieldParts, "."), strings.Join(fieldParts, "."))
+		} else if idx > len(array) {
+
+			if len(array) == 0 {
+				// case (6) or (7)
+				return nil, fmt.Errorf("array indexes must start from 0. No index found before index '%v' in field '%v'",
+					idx, strings.Join(fieldParts, "."))
+			}
+			// case (7) and (8)
+			return nil, fmt.Errorf("field '%v' contains an array index that does not increase sequentially from field '%v.%d'. "+
+				"Array indexes in fields must start from 0 and increase sequentially",
+				strings.Join(fieldParts, "."),
+				strings.Join(fieldParts[:partIndex], "."),
+				(len(array) - 1))
+		}
+
+		array = append(array, true)
+		return array, nil
+	}
+
```

Comment:
At this point, for a non-terminal node, we must check `idx <= len(array)` -- either it exists or it must be the next sequential index.  This checks for the case 'a.0.b, a.2.b'.